### PR TITLE
Improves "yarn workspaces info"

### PR DESCRIPTION
**Pull request by arcanis**
**Summary**

Running `yarn workspaces info` will now yield information to reconstruct the workspace dependency tree. This information can then be feed to other build systems.

**Test plan**

Updated the tests to include the new fields

### CODE REVIEW

File: src/cli/commands/workspaces.js [hunk 1]
```diff
@@ -27,8 +29,30 @@ export async function info(config: Config, reporter: Reporter, flags: Object, ar
   const publicData = {};
 
   for (const workspaceName of Object.keys(workspaces)) {
+    const {loc, manifest} = workspaces[workspaceName];
+
+    const workspaceDependencies = new Set();
+    const mismatchedWorkspaceDependencies = new Set();
+
+    for (const dependencyType of DEPENDENCY_TYPES) {
```
Comment (BYK) [inline src/cli/commands/workspaces.js]: Isn't it better to just omit this before the loop?

File: src/cli/commands/workspaces.js [hunk 2]
```diff
@@ -27,8 +29,30 @@ export async function info(config: Config, reporter: Reporter, flags: Object, ar
...
+      if (dependencyType !== 'peerDependencies') {
+        for (const dependencyName of Object.keys(manifest[dependencyType] || {})) {
+          if (Object.prototype.hasOwnProperty.call(workspaces, dependencyName)) {
```
Comment (BYK) [inline src/cli/commands/workspaces.js]: Don't we have a shortcut for `Object.prototype.hasOwnProperty.call` here? Do we _really_ need it?

File: src/cli/commands/workspaces.js [hunk 3]
```diff
@@ -27,8 +29,30 @@ export async function info(config: Config, reporter: Reporter, flags: Object, ar
...
+            const request = manifest[dependencyType][dependencyName];
```
Comment (BYK) [inline src/cli/commands/workspaces.js]: `expectedRange`, `requestedRange`? Something that wouldn't confuse me with an actual HTTP request or the `request` module?

File: src/cli/commands/workspaces.js (hunk 1 above)
Comment (arcanis) [reply src/cli/commands/workspaces.js]: What do you mean? Duplicate the DEPENDENCY_TYPES array, but without peerDependencies?

File: src/cli/commands/workspaces.js (hunk 2 above)
Comment (arcanis) [reply src/cli/commands/workspaces.js]: I don't think we have, there's multiple matches for this function in the codebase.

Comment (BYK) [reply src/cli/commands/workspaces.js]: Ah, we should make that happen. That said I'm not sure if we need to be so defensive about this. We should be able to use `Object.hasOwnProperty()` I guess? Or some variant of `Object.keys()` that only returns own properties?

File: src/cli/commands/workspaces.js (hunk 1 above)
Comment (BYK) [reply src/cli/commands/workspaces.js]: Yes. It seems wasteful and less readable this way to check for this specific value in every iteration.

File: src/cli/commands/workspaces.js (hunk 2 above)
Comment (arcanis) [reply src/cli/commands/workspaces.js]: `Object.keys` only returns own properties, but we're calling it on a different object from the one we're dereferencing. So for example, a malicious (or a bit weird) user could add a dependency called `__proto__` to one of its workspaces, and without this check we would then access `workspaces.__proto__`.

File: src/cli/commands/workspaces.js (hunk 1 above)
Comment (arcanis) [reply src/cli/commands/workspaces.js]: Yeah, maybe we could have a `DIRECT_DEPENDENCY_TYPES` array.

**Pull request merged**